Add a simple instrumentation framework to Jardin

drdrsh opened the pull request:
This PR adds a simple pub/sub instrumentation framework to Jardin. 

The ideal use case is for applications to consume events and emit metrics/APM traces as necessary.

klazuka commented on jardin/instrumentation/notifier.py:
```diff
@@ -0,0 +1,88 @@
+
+  @classmethod
+  def subscribe(self, handler):
+      if not self.is_enabled():
```
Instead of sprinkling these `is_enabled` checks across the class, what about having an additional no-op implementation? Then use the env var at boot time to pick one implementation.

drdrsh replied:
Where would the "boot time" logic live?

klazuka replied:
In [config.init()](https://github.com/instacart/jardin/blob/86339a14e14c87df72236fcf1a300ac8d95c54bc/jardin/config.py#L23-L44) or at the top-level of the module. I would prefer the former since the initialization order is better defined (although due to a recent change, it can be called in 2 different places 😦 ).

But now that you've made the change to synchronous callbacks (thank you 🙏), we shouldn't need the boolean enabled flag at all.

drdrsh replied:
I am still going to retain the enabled flag in case the subscriber we end up writing is terrible and we want to disable it quickly.

klazuka commented on jardin/instrumentation/notifier.py:
```diff
@@ -0,0 +1,88 @@
+
+  @classmethod
+  def handle_error(self, event_name, subscriber, exc):
+      config.logger.error("Failed to report event {} to {} due to {}".format(event_name, str(type(subscriber)), str(exc)))
```
Use python's new-ish "f-strings". Our min python version is 3.7. 

You also shouldn't need the `str()` calls since string interpolation will call `__format__` which should do the same thing (in most cases). But you can double-check that for the values you want to capture.

klazuka commented on jardin/instrumentation/instrumenter.py:
```diff
@@ -0,0 +1,21 @@
+      self.monotonic_start = time.monotonic()
+      self.start_time = time.time()
+
+  def __exit__(self, exc_type, exc_value, exc_traceback):
```
Probably should differentiate between a call that finished successfully vs with an exception. In the latter case, the duration will be unusually short, skewing your metrics.

Try using the [@contextmanager decorator](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager). It makes these sorts of things a little easier to follow. I always forget how to use `__exit__()` correctly.

drdrsh replied:
I am okay with shorter durations as I plan to use this feature to tag spans and spans with errors are okay to cut short. I agree, though, that the subscriber might be interested in knowing about the failure.

klazuka commented on jardin/instrumentation/notifier.py:
```diff
@@ -0,0 +1,88 @@
+      if not self.is_enabled() or self._should_stop:
+          return
+      try:
+          self._q.put_nowait((event_name, start_time, end_time, duration, tags))
```
This tuple is getting large. Maybe use a [dataclass](https://docs.python.org/3.7/library/dataclasses.html) instead? (we're using min python 3.7 so it's safe).

klazuka commented on jardin/instrumentation/notifier.py:
```diff
@@ -0,0 +1,88 @@
+      while True:
+          try:
+              event_name, start_time, end_time, duration, tags = self._q.get()
+              if event_name is None:
```
could you factor out either a sentinel value? or if you decide to turn your queue items into objects, add a boolean property to indicate whether it's the sentinel/exit signal?

klazuka reviewed (commented):
Nice. Here's my first round of comments. I'll look at the multi-threading in a subsequent round.

klazuka commented on jardin/database/base_client.py:
```diff
@@ -71,8 +76,10 @@ def execute(self, *query, write=False, **kwargs):
-            cursor = self.execute_impl(self._conn, *query)
+                with Instrumenter("connection_initiated", tags=self.tags()):
+                    self._conn = self.connect_impl()
+            with Instrumenter("query", tags=self.tags({"query": query})):
```
Isn't the query text going to be very high cardinality? Or does that not matter for APM?

drdrsh replied:
This will up to the subscriber to handle as they deem fit, for APM, we may want to add the query as a tag for the span, for metrics we probably won't.

klazuka commented on jardin/instrumentation/event.py:
```diff
@@ -0,0 +1,10 @@
+@dataclass
+class Event:
+  name: str
+  start_time: float = 0.0
```
needs a comment: seconds since unix epoch?

klazuka replied:
Also, I would consider factoring out `start_time`, `end_time` and `duration` since they will either all be set or all be unset. Right now it's too easy to create an invalid configuration.

Maybe something like:
```python
@dataclass
class Event:
    name: str
    timing: Optional[EventTiming] = None
    tags: Dict[str, str] = field(default_factory=dict)

@dataclass
class EventTiming:
    start_time: float
    end_time: float
    duration: float
```

klazuka replied:
And one more thing 😄  ...
Couldn't `end_time` be derived by adding `duration` to the `start_time`?

klazuka commented on jardin/instrumentation/event.py:
```diff
@@ -0,0 +1,10 @@
+  name: str
+  start_time: float = 0.0
+  end_time: float = 0.0
+  duration: float = 0.0
```
How is the duration measured? seconds? millis?

drdrsh replied:
fractional seconds. I added it to field name.

klazuka commented on jardin/instrumentation/event.py:
```diff
@@ -0,0 +1,10 @@
+
+@dataclass
+class Event:
+  name: str
```
4 space indents please 😄

klazuka commented on jardin/instrumentation/base_subscriber.py:
```diff
@@ -0,0 +1,8 @@
+from jardin.instrumentation.event import Event
+
+
+class BaseSubscriber:
```
1. Whitespace at the top of the file is weird
2. For a simple callback like this, could we just have the client provide a function that takes an `Event` as a parameter?

drdrsh replied:
For 2, I could go either way. But I think using callbacks might make this less future-proof if we want to add more methods to the subscriber

klazuka commented on jardin/instrumentation/notifier.py:
```diff
@@ -0,0 +1,46 @@
+import os
+import uuid
+import queue
+import atexit
```
some of these imports are no longer being used.

klazuka commented on jardin/instrumentation/notifier.py:
```diff
@@ -0,0 +1,46 @@
+  def report_event(self, ev: Event):
+      subscribers = dict(self._subscribers)
+      for subscriber_id in dict(self._subscribers):
+          subscriber = subscribers[subscriber_id]
```
This is confusing. It looks like you are making a shallow copy to avoid external mutation while looping over the subscribers?

I haven't yet reviewed this PR's thread-safety, but I would consider using a lock.

drdrsh replied:
That is correct. 
I wanted to avoid adding locks since the race condition is fairly benign (we miss notifying a subscriber that has just been added)

Will add the lock just to avoid the confusion.

drdrsh commented:
I now bundle exception information with `connection_banned` and `query_retry` events. Fixed some linting issues, used locks in event fanout, added a null notifier class and placed the class selection logic in `config.py` and added `EventTiming` and `EventExceptionInformation` classes to `Event` object.

klazuka commented on jardin/instrumentation/notifier.py:
```diff
@@ -0,0 +1,45 @@
+
+    def report_event(self, ev: Event) -> None:
+        with self.lock:
+            for subscriber_id in self._subscribers:
```
You can simplify this by just iterating over the values:
```python
for subscriber in self._subscribers.values():
    ...
```

klazuka commented on jardin/instrumentation/notifier.py:
```diff
@@ -0,0 +1,45 @@
+from jardin.instrumentation.base_subscriber import BaseSubscriber
+from jardin.instrumentation.event import Event
+
+class NullNotifier:
```
Proposal for further simplification:

1. delete the `JARDIN_INSTRUMENTATION_ENABLED` env var
2. delete this `NullNotifier` class
3. rename `SimpleNotifier` to `Notifier`
4. In `Notifier.report_event`, without taking the lock, return early if there are zero subscribers

And maybe add a comment to say that we are intentionally not taking the lock because the race condition isn't worth it.

klazuka commented on jardin/instrumentation/notifier.py:
```diff
@@ -0,0 +1,45 @@
+            for subscriber_id in self._subscribers:
+                subscriber = self._subscribers[subscriber_id]
+                try:
+                    subscriber.report_event(ev)
```
Is it worth holding the lock while we make this callback? I don't expect the callers would need it to be serialized, and if they do, they can do the locking themselves.

drdrsh replied:
What's a clean way to do that? 
The one way I can think of is to hold the lock, make a copy of subscribers, release the lock and iterate on that list. 
Releasing the lock before reporting the event and reacquiring it after wards feels a bit awkward + it will allow modification of the subscribers list during iteration.

klazuka replied:
Your way sounds fine (copy the list of subscribers). And, yes, I do realize that we're basically back to what you were doing at the beginning of the PR when I started nit-picking 😄
